### ISSUE TITLE
Improve encoding of grouping sets in plan

### ISSUE THREAD

**Issue by martint** [prestodb/presto#10478]
Currently, the list of concrete grouping sets is expanded during analysis. This results in inefficiencies during analysis and planning due to redundant re-analysis and storage of the atoms of each grouping element.

Instead, the engine should only analyze the atomic units (columns of each grouping element) and should encode the expected grouping sets in the plan using an abstract description of the components instead of enumerating every combination of grouping columns. The enumeration should happen when the operator is instantiated.

I.e., the GroupId node should contain a list of grouping elements:

```
Elements = (CUBE set<symbol> | ROLLUP list<symbol> | SETS list<set<symbol>>)+
```

**Comment by sopel39** [comment]
Does this cause nasty failures of coordinator?

**Comment by martint** [comment]
Yes, it allocates 2^#columns sets of expressions when analyzing the query. It then analyzes each of those expressions and creates a GroupID node with that many entries.

**Comment by nezihyigitbasi** [comment]
Fixed by #10477.

**Comment by martint** [comment]
This is still not fixed. #10477 is a workaround for the immediate issue of queries not even making it to the point where the QueryStarted event is submitted.
